### PR TITLE
Normalize the non-deterministic `Tasks completed` output in the `ReduceFirstValue` docs example

### LINKED ISSUE #6663: `docs/graph/builder/joins.md` `first_value_reducer` example asserts a timing-dependent `Tasks completed: 1` that flakes CI
This issue was posted by Claude Code using claude-opus-4-8 on behalf of David. David approved filing this issue conditional on a reliable trigger existing; he has not reviewed the wording line by line.

### Summary

The graph suite already treats "how many branches completed under `ReduceFirstValue`" as an unsafe thing to assert: `test_reduce_first_value` in `tests/graph/builder/test_joins_and_reducers.py` deliberately checks only that *a* result came back and that the winner ran, with the comment "timing can be tricky, so we just verify we got a result." The `first_value_reducer.py` example in `docs/graph/builder/joins.md` is the outlier: it hard-asserts `#> Tasks completed: 1`. That count is not guaranteed, so the example flakes under CI load. The returned value (`Result from task 1`) is stable; only the side-effect count is not.

### Observed failure

Seen while triaging CI on a different PR (#6592), on the `test on 3.10 (all-extras)` shard, run https://github.com/pydantic/pydantic-ai/actions/runs/29927438916/job/88948129904 :

```
tests/test_examples.py::test_docs_examples[docs/graph/builder/joins.md:291:first_value_reducer.py] - Failed: Print output changed code:
       #> Result from task 1
       print(f'Tasks completed: {state.tasks_completed}')
  -    #> Tasks completed: 1
  +    #> Tasks completed: 2
```

The example only runs on the all-extras / lowest-versions shards, which run under `coverage run -m pytest -n logical` (coverage tracing plus one xdist worker per logical CPU), so the event loop gets starved.

### Why the count is not a safe assertion

`ReduceFirstValue.__call__` (`pydantic_graph/pydantic_graph/join.py`) only sets a cooperative flag via `ctx.cancel_sibling_tasks()`. The executor (`_GraphIterator` in `pydantic_graph/pydantic_graph/graph_builder.py`) acts on that flag only when it processes the winner's result: it reduces the winning value, then dispatches `_cancel_sibling_tasks`. A losing branch whose own work reaches its side effect (`state.tasks_completed += 1`) before that dispatch keeps it. Under a starved shard, a losing branch's `asyncio.sleep` timer can fire in exactly that window, so the count comes out as 2 or more. This is inherent to cooperative cancellation: `ReduceFirstValue` guarantees the winning value and early-stops branches that are still running, but it cannot roll back a side effect a branch already executed.

So the defect is in the example (an over-specified assertion), not in the runtime.

### Reliable reproduction: what is and is not deterministic

This distinction matters for how a fix can be tested, so stating it plainly:

- The wall-clock race that actually fired in CI (a *sleeping* sibling's timer firing before cancellation is dispatched) is **not** deterministically reproducible without patching internals. The executor dispatches cancellation synchronously right after the reduce, with no `await` in between, so in single-threaded cooperative scheduling a suspended sibling can only "beat" the cancel by scheduling luck. Reproducing the race on demand requires injecting artificial lag into `_GraphIterator._cancel_sibling_tasks`, which is not a basis for a committed test.
- The underlying semantic boundary **is** deterministically reproducible through the public `GraphBuilder` API, with no internal patching: a branch that reaches its side effect before cancellation is dispatched keeps it, while a branch still suspended before its side-effect line is cancelled cleanly. Constructed structurally (a loser with no `await` in front of its side effect, plus a sibling parked on an `asyncio.Event` that is never set), this holds across thousands of runs including under event-loop load.

The practical consequence: a committed regression test can pin the contract the example violates, but it does **not** re-stage the flaky assertion. There is no test that reproduces the wall-clock flake itself, and none is proposed, because that cannot be done honestly without patching internals. The regression test protects the contract (`ReduceFirstValue` keeps an already-run side effect; it is not side-effect-exclusive), which is what makes the example's `Tasks completed: 1` wrong.

### Proposed fix

- In the example, remove the timing-dependent `Tasks completed: 1` assertion. The winning value `Result from task 1` is the real guarantee and stays.
- Add a regression test next to `test_reduce_first_value` that deterministically pins the semantic boundary described above.

A PR follows.

### PR DESCRIPTION
**Pull request by dsfaccini**
This pull request was posted by Claude Code using claude-opus-4-8 on behalf of David. David reversed his own earlier close of this PR after I surfaced that the docs-examples harness already normalizes non-deterministic printed output; he approved this approach and reusing the branch, but has not reviewed the diff line by line.

- Closes #6663

The `first_value_reducer.py` example in `docs/graph/builder/joins.md` printed `#> Tasks completed: 1`, a timing-dependent side-effect count that flaked CI under load (`Tasks completed: 2`). `ReduceFirstValue` guarantees the winning value and early-stops branches that are still running, but it cannot roll back a side effect a branch already executed, so the completed count is not deterministic.

Rather than reduce the example to remove the count, this PR keeps it intact (state, counter, and the lesson that siblings are cancelled) and normalizes the non-deterministic value the same way the harness already normalizes `datetime`, `run_id`, and `conversation_id`:

- `tests/test_examples.py` `print_callback` now rewrites `Tasks completed: <n>` to `Tasks completed: ...` before the printed output is compared, so `run_print_check` matches regardless of how many siblings finished. The `#>` line in the example becomes `#> Tasks completed: ...`, which is also the accurate rendering of a value `ReduceFirstValue` does not make deterministic (the printed `1` was wrong under load).
- `test_reduce_first_value_keeps_completed_siblings_side_effects` (next to `test_reduce_first_value`) pins the semantic boundary deterministically: branches that reach their side effect before cancellation is dispatched keep it, while a branch still suspended before its side-effect line is cancelled cleanly.

I verified the normalization on the real comparison path with a forced-divergent count: injecting lag into sibling-cancellation dispatch drives the example's count to 4, and `run_print_check` still passes with the normalization (and correctly fails, showing `Tasks completed: 4`, without it).

On what the regression test does and does not cover: the wall-clock race that actually fired in CI (a sleeping sibling's timer firing before cancellation is dispatched) is not deterministically reproducible without patching internals, so no test re-stages that flake and none claims to. The test uses only the public `GraphBuilder` API to pin the contract the printed `1` violated (`ReduceFirstValue` is not side-effect-exclusive); it is structurally deterministic and holds under event-loop load. The flake itself is handled by the normalization above, not by the test.

No runtime or public-API changes. The example's code is unchanged from `main` except the `#>` output line. Targeted tests pass (the joins examples, the new test, and the neighboring `test_reduce_first_value`); coverage is intact.

### Checklist

- [ ] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [x] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [x] **PR title** is fit for the [release changelog](https://github.com/pydantic/pydantic-ai/releases).

### CODE REVIEW

**Comment by dsfaccini** [comment]
This comment was posted by Claude Code using claude-opus-4-8 on behalf of David.

## Review beacon — verdict: PASS (updated after approach pivot)

**Update:** this PR was closed and reopened. The first approach (remove the `Tasks completed` assertion and the orphaned counter) was closed as net-negative because it reduced the example's teaching value for a rare flake. The reopened PR takes a non-reductive approach: the docs-examples harness already normalizes non-deterministic printed output via `print_callback`, so the flake is killed by normalizing `Tasks completed: <n>` to `Tasks completed: ...` (as it already does for `datetime` / `run_id` / `conversation_id`) and rendering the example's `#>` line accordingly. The example's code is unchanged from `main` except that one output line.

### Current diff

- `tests/test_examples.py` `print_callback` normalizes the non-deterministic count.
- `docs/graph/builder/joins.md` `#>` line reads `Tasks completed: ...`.
- `tests/graph/builder/test_joins_and_reducers.py` keeps `test_reduce_first_value_keeps_completed_siblings_side_effects`, which pins the semantic boundary (unchanged from the prior round).

### Verified

- Normalization applies on the real `run_print_check` comparison path. With sibling-cancellation dispatch delayed so the example's count diverges to 4, `run_print_check` still passes with the normalization, and correctly fails (showing `Tasks completed: 4`) without it. So the fix is load-bearing on the actual path, not on the count happening to be 1.
- The regression test genuinely exercises `ReduceFirstValue` cancellation (a broken cancel would hang) and is structurally deterministic under load.
- Targeted tests pass (joins examples, the new test, `test_reduce_first_value`); coverage intact; `uv.lock` not in the diff.

Lower-risk than the closed approach: no example content removed, one normalization line plus one doc-output line, same regression test.

File: tests/graph/builder/test_joins_and_reducers.py [hunk 1]
```diff
@@ -370,3 +370,55 @@ async def slow_process(ctx: StepContext[StateWithResults, None, int]) -> str:
...
     assert 'completed-1' in state.results
+
+
+async def test_reduce_first_value_keeps_completed_siblings_side_effects():
+    """`ReduceFirstValue` guarantees the winning value, not side-effect exclusivity.
+
+    It early-stops siblings that are still suspended, but it cannot un-run a side effect a
+    branch already executed: sibling cancellation is dispatched only once the winner's value
+    is reduced, so any sibling that reached its side effect before then keeps it. Branches 1
+    and 2 record before cancellation is dispatched (nothing awaits in front of the side effect)
```
[the remark this answers is not in the document]
Comment (dsfaccini) [reply tests/graph/builder/test_joins_and_reducers.py]: This reply was posted by Claude Code using claude-opus-4-8 on behalf of David (who directed the fix but did not line-review the new test).

Good catch — you were right. The old assertion `state.completed == ['fast-1', 'fast-2']` only pinned "branches that ran to completion kept their effects," which is trivially true: branches 1 and 2 had no `await` before their side effect, so they were never cancellation candidates, and the one suspended branch was cancelled *before* its side-effect line. Nothing exercised the documented boundary.

Fixed in d3bd10a07 by rewriting the test around three `asyncio.Event`-gated branches so the ordering is causal, not timed:

- `winner` awaits `effect_recorded` before reducing, so cancellation is dispatched only once the losing sibling has written and parked;
- `effect-then-suspend` records its side effect, then suspends on `never` → cancelled *after* the write → effect survives;
- `suspend-then-effect` suspends *before* its side-effect line → cancelled without recording.

`assert state.completed == ['effect-then-suspend', 'winner']` now pins both halves of the guarantee. Verified 500/500 non-flaky in one process, and discriminating: commenting out `ctx.cancel_sibling_tasks()` makes the run hang (the suspended siblings are never freed), so the surviving-effect assertion is reachable only because cancellation fires.
